<a href="https://colab.research.google.com/github/guy998877/projects/blob/main/Assignment3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
pd.options.mode.chained_assignment = None  # default='warn'

# Part 1: The Marvel Universe Social Network

In [ ]:
edges = pd.read_csv("edges.csv")
hero_network = pd.read_csv("hero-network.csv")
nodes = pd.read_csv("nodes.csv")

#### **Answer 1:**
in the table hero_network we have all the social connections between heroes.

In [ ]:
hero_network.describe()

hero1            hero2
count            574467           574467
unique             6211             6173
top     CAPTAIN AMERICA  CAPTAIN AMERICA
freq               8149             8350

We need to
* remove rows when a hero is a friend of itself
* remove duplicates

to prevent bugs.

In [ ]:
hero_network.drop_duplicates()
hero_network.drop(hero_network[hero_network['hero2'] == hero_network['hero1']].index, inplace=True)
hero_relation_count = hero_network.groupby(['hero1']).count()

All the heroes that have friends are:

In [ ]:
hero_relation_count[hero_relation_count['hero2'] == 1]

hero2
hero1                      
AFTERLIFE                 1
AGAMEMNON II/ANDREI       1
AGENT-X/                  1
AGGAMON                   1
ALANYA                    1
...                     ...
YEE, WAI CHEE             1
YLA                       1
YOUNG, AGNES              1
ZABO                      1
ZONE                      1

[265 rows x 1 columns]

#### **Answer 2**

In [ ]:
captain = hero_network.loc[hero_network['hero1'].str.contains("CAPTAIN AMERICA")].drop_duplicates() # friends of captain america
spiderman = hero_network.loc[hero_network['hero1'].str.contains("SPIDER-MAN")].drop_duplicates() # friends of spiderman
ironman = hero_network.loc[hero_network['hero1'].str.contains("IRON MAN")].drop_duplicates() # friends of ironman

In [ ]:
captain_spiderman = captain[captain['hero2'].isin(spiderman['hero2'])].drop_duplicates() # friends of captain america and spiderman
all_mutual = captain_spiderman[captain_spiderman['hero2'].isin(ironman['hero2'])].drop_duplicates() # friends of all 3 heroes

In [ ]:
all_mutual.describe()

hero1                 hero2
count               558                   558
unique                6                   473
top     CAPTAIN AMERICA  INVISIBLE WOMAN/SUE 
freq                472                     3

We can see that there are 473 unique values at hero2 column so Captain America, Spiderman and Ironman have **473 unique** mutual friends

#### **Answer 3**
Information on all heros is in nodes table

In [ ]:
nodes.groupby(nodes['node'].str.len()).max()

node  type
node                                  
1                              T  hero
2                             TU  hero
3                            ZON  hero
4                           ZURI  hero
5                          ZZZAX  hero
6                         ZAWADI  hero
7                        ZEAKLAR  hero
8                       ZEY-ROGG  hero
9                      ZIRCONIUM  hero
10                    ZELIG, LON  hero
11                   rse powers.  hero
12                  ZAMORA, PETE  hero
13                 ZEALOT/MOREAU  hero
14                ZOLA, ARNIM II  hero
15               ZEDRAO, GENERAL  hero
16              ZOOG [KYMELLIAN]  hero
17             ZEITGEIST, NICOLA  hero
18            ZOMBIE/SIMON GARTH  hero
19           cept. This listing,  hero
20          ZODIAK/NORMAN HARRIS  hero
26    SPIDER-MAN/PETER PARKERKER  hero

It is clear that the maximum length of name is 26 characters. let's find out if there are more heros with the same name length.

In [ ]:
nodes[nodes['node'].str.len() == 26]

node  type
14618  SPIDER-MAN/PETER PARKERKER  hero

As seen, there is only 1 hero that has the most characters in his name and it's 'SPIDER-MAN/PETER PARKERKER'

In [ ]:
nodes.groupby(nodes['node'].str.len()).min()

node   type
node                                   
1                              A   hero
2                             IT  comic
3                            A 1  comic
4                           A 10  comic
5                          A '00  comic
6                         2001 8  comic
7                        2001 10  comic
8                       AA2 10/2  comic
9                      ADVERSARY  comic
10                    ADAM 3,031  comic
11                   ABOMINATRIX  comic
12                  A:TERM OBJ 2  comic
13                 A: TERM OBJ 2  comic
14                ADAMSON, JASON  comic
15               AGUIRRE, ISOBEL  comic
16              4-D MAN/MERCURIO  comic
17             ADMIRAL PROTOCOL/  comic
18            ACHILLES II/HELMUT  comic
19           ADAMS, NICOLE NIKKI  comic
20          24-HOUR MAN/EMMANUEL  comic
26    SPIDER-MAN/PETER PARKERKER   hero

In [ ]:
nodes[nodes['node'].str.len() == 1]

node  type
738      A  hero
10274    M  hero
12225    P  hero
13191    Q  hero
16269    T  hero

So it's clear that there are 5 heros with the minimum name length and they are: 'A', 'M', 'P', 'Q', 'T'

#### **Answer 4**

In [ ]:
def not_mutual_with_cap(hero_name):
  cap_friends = hero_network[hero_network['hero1'].str.contains("CAPTAIN AMERICA")]
  hero_friends = hero_network[hero_network['hero1'].str.contains(hero_name)]

  return hero_friends[np.logical_not(hero_friends['hero2'].isin(cap_friends['hero2']))]

In [ ]:
not_mutual_with_cap("SPIDER-MAN").describe()

hero1        hero2
count                   2255         2255
unique                     3          738
top     SPIDER-MAN/PETER PAR  STACY, GWEN
freq                    1838           50

We can see that spiderman has **738 unique** friends that are not friends of captain america.

#### **Answer 5**

In [ ]:
friends_of_spiderman = hero_network[hero_network['hero1'].isin(spiderman['hero2'])].groupby(['hero1']).count().drop_duplicates()
friends_of_spiderman.sum()

hero2    351992
dtype: int64

So we found out that all friends of spiderman has a total of 351,992 friends.

#Part 2: Earthquake

In [ ]:
df= pd.read_csv("database.csv")

NameError: ignored

In [ ]:
df

Date      Time  ...  Magnitude Source     Status
0      01/02/1965  13:44:18  ...            ISCGEM  Automatic
1      01/04/1965  11:29:49  ...            ISCGEM  Automatic
2      01/05/1965  18:05:58  ...            ISCGEM  Automatic
3      01/08/1965  18:49:43  ...            ISCGEM  Automatic
4      01/09/1965  13:32:50  ...            ISCGEM  Automatic
...           ...       ...  ...               ...        ...
23407  12/28/2016  08:22:12  ...                NN   Reviewed
23408  12/28/2016  09:13:47  ...                NN   Reviewed
23409  12/28/2016  12:38:51  ...                US   Reviewed
23410  12/29/2016  22:30:19  ...                US   Reviewed
23411  12/30/2016  20:08:28  ...                US   Reviewed

[23412 rows x 21 columns]

#### **Answer 1**


In [ ]:
df[df['Magnitude']==df['Magnitude'].max()]

Date      Time  ...  Magnitude Source    Status
17083  12/26/2004  00:58:53  ...          OFFICIAL  Reviewed
20501  03/11/2011  05:46:24  ...          OFFICIAL  Reviewed

[2 rows x 21 columns]

#### **Answer 2**

In [ ]:
pd_nuclear = df[df["Type"] == "Nuclear Explosion"]
medians = pd_nuclear["Magnitude"].median()
medians

5.9

#### **Answer 3A**

We sort the table to show earthquake over 5 in magnitude : 

In [ ]:
df_above_five_max = df[(df['Magnitude']) > 5 ]
df_above_five_max

Date      Time  Latitude  ...  Magnitude Source     Status  Year
0      01/02/1965  13:44:18   19.2460  ...            ISCGEM  Automatic  1965
1      01/04/1965  11:29:49    1.8630  ...            ISCGEM  Automatic  1965
2      01/05/1965  18:05:58  -20.5790  ...            ISCGEM  Automatic  1965
3      01/08/1965  18:49:43  -59.0760  ...            ISCGEM  Automatic  1965
4      01/09/1965  13:32:50   11.9380  ...            ISCGEM  Automatic  1965
...           ...       ...       ...  ...               ...        ...   ...
23407  12/28/2016  08:22:12   38.3917  ...                NN   Reviewed  2016
23408  12/28/2016  09:13:47   38.3777  ...                NN   Reviewed  2016
23409  12/28/2016  12:38:51   36.9179  ...                US   Reviewed  2016
23410  12/29/2016  22:30:19   -9.0283  ...                US   Reviewed  2016
23411  12/30/2016  20:08:28   37.3973  ...                US   Reviewed  2016

[23412 rows x 22 columns]

In [ ]:
def find_year(date):
    if date != None:
      list_date = date.split("/")
      year = list_date[-1]
      return year

df_above_five_max['Year'] = df_above_five_max['Date'].apply(find_year)
df_above_five_max

Date      Time  Latitude  ...  Magnitude Source     Status  Year
0      01/02/1965  13:44:18   19.2460  ...            ISCGEM  Automatic  1965
1      01/04/1965  11:29:49    1.8630  ...            ISCGEM  Automatic  1965
2      01/05/1965  18:05:58  -20.5790  ...            ISCGEM  Automatic  1965
3      01/08/1965  18:49:43  -59.0760  ...            ISCGEM  Automatic  1965
4      01/09/1965  13:32:50   11.9380  ...            ISCGEM  Automatic  1965
...           ...       ...       ...  ...               ...        ...   ...
23407  12/28/2016  08:22:12   38.3917  ...                NN   Reviewed  2016
23408  12/28/2016  09:13:47   38.3777  ...                NN   Reviewed  2016
23409  12/28/2016  12:38:51   36.9179  ...                US   Reviewed  2016
23410  12/29/2016  22:30:19   -9.0283  ...                US   Reviewed  2016
23411  12/30/2016  20:08:28   37.3973  ...                US   Reviewed  2016

[23412 rows x 22 columns]

We created a new column "Year" that contains only the year label from "Date" column DD/MM/YYYY with a function for every row in date. 

Next, we used groupby in order to show how many earthquake per year

In [ ]:
events_per_year = df_above_five_max.groupby("Year",as_index=False)["Date"].count()
events_per_year = events_per_year.rename(columns={'Date': 'Count'})
events_per_year

Year  Count
0                       1965    339
1                       1966    234
2                       1967    255
3                       1968    305
4                       1969    323
5                       1970    345
6                       1971    386
7                       1972    388
8                       1973    401
9                       1974    361
10                      1975    411
11  1975-02-23T02:58:41.000Z      1
12                      1976    457
13                      1977    425
14                      1978    410
15                      1979    356
16                      1980    348
17                      1981    321
18                      1982    346
19                      1983    453
20                      1984    482
21                      1985    475
22  1985-04-28T02:53:41.530Z      1
23                      1986    485
24                      1987    505
25                      1988    489
26                      1989    480
27                      1990    528
28                      1991    429
29                      1992    533
30                      1993    466
31                      1994    508
32                      1995    591
33                      1996    541
34                      1997    456
35                      1998    388
36                      1999    446
37                      2000    553
38                      2001    443
39                      2002    444
40                      2003    485
41                      2004    571
42                      2005    533
43                      2006    508
44                      2007    608
45                      2008    508
46                      2009    517
47                      2010    560
48                      2011    712
49  2011-03-13T02:23:34.520Z      1
50                      2012    445
51                      2013    461
52                      2014    480
53                      2015    446
54                      2016    469

After it , we will find the max value in the table

In [ ]:
events_per_year[events_per_year['Count']==events_per_year['Count'].max()]

Year  Count
48  2011    712

#### **Answer 3B**

We will sort the table for eartquake over 8 in magnitude

In [ ]:
earthquake_table = df[(df['Magnitude']) > 8 ]
earthquake_table

Date      Time  Latitude  ...  Magnitude Source     Status  Year
11     01/24/1965  00:11:17   -2.6080  ...            ISCGEM  Automatic  1965
16     02/04/1965  05:01:22   51.2510  ...          OFFICIAL  Automatic  1965
538    10/17/1966  21:42:00  -10.6650  ...            ISCGEM  Automatic  1966
911    05/16/1968  00:49:02   40.8600  ...            ISCGEM  Automatic  1968
2008   07/26/1971  01:23:22   -4.8170  ...            ISCGEM  Automatic  1971
9484   05/23/1989  10:54:46  -52.3410  ...               PAR   Reviewed  1989
11959  06/09/1994  00:33:16  -13.8410  ...               HRV   Reviewed  1994
12119  10/04/1994  13:22:56   43.7730  ...               HRV   Reviewed  1994
12892  02/17/1996  05:59:31   -0.8910  ...               HRV   Reviewed  1996
13925  03/25/1998  03:12:25  -62.8770  ...               HRV   Reviewed  1998
15440  06/23/2001  20:33:14  -16.2650  ...          OFFICIAL   Reviewed  2001
16446  09/25/2003  19:50:06   41.8150  ...               HRV   Reviewed  2003
17080  12/23/2004  14:59:04  -49.3120  ...               HRV   Reviewed  2004
17083  12/26/2004  00:58:53    3.2950  ...          OFFICIAL   Reviewed  2004
17329  03/28/2005  16:09:37    2.0850  ...          OFFICIAL   Reviewed  2005
18111  11/15/2006  11:14:14   46.5920  ...              GCMT   Reviewed  2006
18211  01/13/2007  04:23:21   46.2430  ...              GCMT   Reviewed  2007
18346  04/01/2007  20:39:59   -8.4660  ...              GCMT   Reviewed  2007
18615  09/12/2007  11:10:27   -4.4380  ...          OFFICIAL   Reviewed  2007
19661  09/29/2009  17:48:11  -15.4890  ...              GCMT   Reviewed  2009
19928  02/27/2010  06:34:12  -36.1220  ...          OFFICIAL   Reviewed  2010
20501  03/11/2011  05:46:24   38.2970  ...          OFFICIAL   Reviewed  2011
21219  04/11/2012  08:38:37    2.3270  ...          OFFICIAL   Reviewed  2012
21224  04/11/2012  10:43:11    0.8020  ...                US   Reviewed  2012
21765  05/24/2013  05:44:49   54.8920  ...                US   Reviewed  2013
22120  04/01/2014  23:46:47  -19.6097  ...                US   Reviewed  2014
22791  09/16/2015  22:54:33  -31.5729  ...                US   Reviewed  2015

[27 rows x 22 columns]

Like the previous question , we will add a column year from column Date

In [ ]:
def find_year(date):
    if date != None:
      list_date = date.split("/")
      year = list_date[-1]
      return year

earthquake_table['Year'] = earthquake_table['Date'].apply(find_year)
earthquake_table

Date      Time  Latitude  ...  Magnitude Source     Status  Year
11     01/24/1965  00:11:17   -2.6080  ...            ISCGEM  Automatic  1965
16     02/04/1965  05:01:22   51.2510  ...          OFFICIAL  Automatic  1965
538    10/17/1966  21:42:00  -10.6650  ...            ISCGEM  Automatic  1966
911    05/16/1968  00:49:02   40.8600  ...            ISCGEM  Automatic  1968
2008   07/26/1971  01:23:22   -4.8170  ...            ISCGEM  Automatic  1971
9484   05/23/1989  10:54:46  -52.3410  ...               PAR   Reviewed  1989
11959  06/09/1994  00:33:16  -13.8410  ...               HRV   Reviewed  1994
12119  10/04/1994  13:22:56   43.7730  ...               HRV   Reviewed  1994
12892  02/17/1996  05:59:31   -0.8910  ...               HRV   Reviewed  1996
13925  03/25/1998  03:12:25  -62.8770  ...               HRV   Reviewed  1998
15440  06/23/2001  20:33:14  -16.2650  ...          OFFICIAL   Reviewed  2001
16446  09/25/2003  19:50:06   41.8150  ...               HRV   Reviewed  2003
17080  12/23/2004  14:59:04  -49.3120  ...               HRV   Reviewed  2004
17083  12/26/2004  00:58:53    3.2950  ...          OFFICIAL   Reviewed  2004
17329  03/28/2005  16:09:37    2.0850  ...          OFFICIAL   Reviewed  2005
18111  11/15/2006  11:14:14   46.5920  ...              GCMT   Reviewed  2006
18211  01/13/2007  04:23:21   46.2430  ...              GCMT   Reviewed  2007
18346  04/01/2007  20:39:59   -8.4660  ...              GCMT   Reviewed  2007
18615  09/12/2007  11:10:27   -4.4380  ...          OFFICIAL   Reviewed  2007
19661  09/29/2009  17:48:11  -15.4890  ...              GCMT   Reviewed  2009
19928  02/27/2010  06:34:12  -36.1220  ...          OFFICIAL   Reviewed  2010
20501  03/11/2011  05:46:24   38.2970  ...          OFFICIAL   Reviewed  2011
21219  04/11/2012  08:38:37    2.3270  ...          OFFICIAL   Reviewed  2012
21224  04/11/2012  10:43:11    0.8020  ...                US   Reviewed  2012
21765  05/24/2013  05:44:49   54.8920  ...                US   Reviewed  2013
22120  04/01/2014  23:46:47  -19.6097  ...                US   Reviewed  2014
22791  09/16/2015  22:54:33  -31.5729  ...                US   Reviewed  2015

[27 rows x 22 columns]

We will use grouby to create a new table for count earthquake per year

In [ ]:
count_events = earthquake_table.groupby("Year",as_index=False)["Date"].count()
count_events = count_events.rename(columns={'Date': 'Count'})
count_events

Year  Count
0   1965      2
1   1966      1
2   1968      1
3   1971      1
4   1989      1
5   1994      2
6   1996      1
7   1998      1
8   2001      1
9   2003      1
10  2004      2
11  2005      1
12  2006      1
13  2007      3
14  2009      1
15  2010      1
16  2011      1
17  2012      2
18  2013      1
19  2014      1
20  2015      1

In [ ]:
count_events[count_events['Count']==count_events['Count'].max()]

Year  Count
13  2007      3

#### **Answer 4**

We will create a new column hour from "Time" column

In [ ]:
def find_hour(time):
    if time != None:
      list_time = time.split(":")
      hour = list_time[0]
      return hour

df['Hour'] = df['Time'].apply(find_hour)
df

Date      Time  Latitude  ...  Magnitude Source     Status  Hour
0      01/02/1965  13:44:18   19.2460  ...            ISCGEM  Automatic    13
1      01/04/1965  11:29:49    1.8630  ...            ISCGEM  Automatic    11
2      01/05/1965  18:05:58  -20.5790  ...            ISCGEM  Automatic    18
3      01/08/1965  18:49:43  -59.0760  ...            ISCGEM  Automatic    18
4      01/09/1965  13:32:50   11.9380  ...            ISCGEM  Automatic    13
...           ...       ...       ...  ...               ...        ...   ...
23407  12/28/2016  08:22:12   38.3917  ...                NN   Reviewed    08
23408  12/28/2016  09:13:47   38.3777  ...                NN   Reviewed    09
23409  12/28/2016  12:38:51   36.9179  ...                US   Reviewed    12
23410  12/29/2016  22:30:19   -9.0283  ...                US   Reviewed    22
23411  12/30/2016  20:08:28   37.3973  ...                US   Reviewed    20

[23412 rows x 22 columns]

We will use groupby in order to create a table for count events per hour, and find the max. 

In [ ]:
g = df.groupby("Hour",as_index=False)["Time"].count()
g = g.rename(columns={'Time': 'Count'})
g[g['Count']==g['Count'].max()]

Hour  Count
13   13   1058

# Part 3: FIFA 2019

In [ ]:
fifa = pd.read_csv('data.csv')

In [ ]:
fifa

Unnamed: 0      ID  ... GKReflexes  Release Clause
0               0  158023  ...        8.0         €226.5M
1               1   20801  ...       11.0         €127.1M
2               2  190871  ...       11.0         €228.1M
3               3  193080  ...       94.0         €138.6M
4               4  192985  ...       13.0         €196.4M
...           ...     ...  ...        ...             ...
12482       12482  239724  ...        6.0           €421K
12483       12483  244588  ...       12.0           €1.6M
12484       12484  197998  ...       63.0           €431K
12485       12485  229998  ...       12.0           €743K
12486       12486  238446  ...        NaN             NaN

[12487 rows x 89 columns]

#### **Answer 1**

In [ ]:
def strongest_player(nation):
  nation_players = fifa[fifa['Nationality'] == nation]
  max_overall = max(nation_players['Overall'])
  return nation_players[nation_players['Overall'] == max_overall]['Name']

In [ ]:
print("Strongest player in Brazil:")
print(strongest_player('Brazil'))
print("-------------------------------")
print("Strongest player in France:")
print(strongest_player('France'))

Strongest player in Brazil:
2    Neymar Jr
Name: Name, dtype: object
-------------------------------
Strongest player in France:
14        N. Kanté
17    A. Griezmann
Name: Name, dtype: object


* The strongest player in Brazil is "Neyman Jr"
* The strongest players in France are "N. Kanté" and "A. Griezmann" (Both have the same and highest Overall rating)

#### **Answer 2**

In [ ]:
def club_overall(clubname):
  club_players = fifa[fifa['Club'] == clubname]
  return club_players['Overall'].mean()